OpenTrons/opentrons-api
# feature(api): Allow loading a Magnetic block as a module in PAPI

# Overview

closes https://opentrons.atlassian.net/browse/RLAB-243.
Allow loading a magnetic block as a module.

# Test Plan

- Upload the following python protocol - this should pass analysis and run should be successful.
```
from typing import cast
from opentrons import protocol_api, types

metadata = {
    "protocolName": "Mad Block",
    "author": "Opentrons <engineering@opentrons.com>",
    "description": "A Magnetic Module Test",
    "source": "Opentrons Repository",
    "apiLevel": "2.15",
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    mm = cast(protocol_api.MagneticBlockContext, ctx.load_module("magneticBlockV1", 3))
    labware = mm.load_labware("nest_96_wellplate_100ul_pcr_full_skirt")
```
- add a load_instrument and move to labware, make sure it all works ok.

- Upload the following python protocol - should raise an exception that the module is not supported.
```
from typing import cast
from opentrons import protocol_api, types

metadata = {
    "protocolName": "Mad Block",
    "author": "Opentrons <engineering@opentrons.com>",
    "description": "A Magnetic Module Test",
    "source": "Opentrons Repository",
    "apiLevel": "2.14",
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    mm = cast(protocol_api.MagneticBlockContext, ctx.load_module("magneticBlockV1", 3))
    labware = mm.load_labware("nest_96_wellplate_100ul_pcr_full_skirt")
```

- Try to access module context serial_number property for all modules. should be accessible for all modules besides the Magnetic block.


# Changelog

- Added `NonConnectedModuleCore` that inherits from AbstractModuleCore and changed locations that use `ModuleCore` to use both `ModuleCore` and `NonConnectedModuleCore`.
- Moved `get_serial_number` from the `AbstractModuleCore` to the derived classes except `AbstractMagneticModuleCore`.
- Changed core/engine/protocol `_module_cores_by_id` to type `MutableMapping` instead of `Dict` to support a Union value type.
- Moved `serial_number` property in the public context to all module context besides Mag block.
- Added condition in `load_module` public context to not allow Mag block module in versions less than 2.14

# Review requests

Changes make sense?
Did I miss anything?

# Risk assessment

Medium. Need to smoke test ModuleContext for all modules. need to practice loading module on labware and moving labware from module. Make sure `serial_number` is available only for non mag block modules.
